File: src/main/scala-2.11/com/lei/leetcode/P154/Solution.py
class Solution(object):
    def findMin(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """

        if nums is None:
            return
        if len(nums) == 1:
            return nums[0]
        if len(nums) == 2:
            return min(nums[0], nums[1])
        lo = 0
        hi = len(nums) - 1
        if nums[hi] > nums[lo]:
            return nums[lo]

        mid = lo + (hi - lo) / 2
        left = nums[lo] if (nums[lo] < nums[mid]) else self.findMin(nums[lo:mid])
        right = nums[mid] if (nums[mid] < nums[hi]) else self.findMin(nums[mid:hi + 1])
        return min(left, right)


if __name__ == '__main__':
    print Solution().findMin([1, 3, 3])
    print Solution().findMin([3, 3, 1, 3])
    print Solution().findMin([3, 4, 4, 4, 4, 4, 4, 4, 5, 6, 2])
    print Solution().findMin([10, 1, 10, 10, 10])

File: src/main/scala-2.11/com/lei/leetcode/P154/Solution.java
package com.lei.leetcode.P154;

/**
 * 1. remove the equal value on both side.
 * 2. then it becomes the linear monotonically increasing problem. (may have the equal value in the middle, but not a problem)
 */
public class Solution {
    public int findMin(int[] nums) {
        if (nums == null || nums.length == 0) {
            return Integer.MIN_VALUE;
        }

        int start = 0, end = nums.length - 1;

        //only need to add the following while loop on top of the solution
        //for Part I
        //if two line segments have overlap, remove the overlap.
        //so, the problem can be solved as Part I
        while (nums[end] == nums[start] && end > start) {
            end--;
        }

        while (start < end) {
            //if the linear monotonically increasing in [start, end]
            if (nums[start] < nums[end]) {
                return nums[start];
            }

            int mid = start + (end - start) / 2;
            if (nums[mid] >= nums[start]) {
                start = mid + 1;
            } else {
                end = mid;
            }
        }

        return nums[start];
    }
}

